kyo-core/shared/src/main/scala/kyo/scheduler/IOTask.scala

```diff
@@ -36,12 +36,12 @@ sealed private[kyo] class IOTask[Ctx, E, A] private (
 
     private inline def locally[A](inline f: A): A = f
 
-    final private def eval(startMillis: Long, clock: InternalClock)(using Safepoint) =
+    final private def eval(startMillis: Long, clock: InternalClock, deadline: Long)(using Safepoint) =
         try
             curr = Isolate.restoring(trace, this) {
                 ArrowEffect.handlePartial(erasedAbortTag, Tag[Async.Join], curr, context)(
                     stop =
-                        shouldPreempt(),
+                        shouldPreempt() || (deadline != Long.MaxValue && clock.currentMillis() > deadline),
```

Comment:
This shouldn't introduce overhead for JVM but it can be costly for JS. I think we can optimize it later if needed.